src/thorin/transform/cleanup_world.cpp

```diff
@@ -233,19 +232,20 @@ next_continuation:;
 
 void Cleaner::rebuild() {
     Importer importer(world_);
-    importer.type_old2new_.rehash(world_.types().capacity());
-    importer.def_old2new_.rehash(world_.defs().capacity());
 
-    for (auto&& [_, cont] : world().externals()) {
+    for (auto&& [_, cont] : world_.externals()) {
         if (cont->is_exported())
             importer.import(cont);
     }
 
-    swap(importer.world(), world_);
+    for (auto [cont, func] : world_.data_.plugin_intrinsics_)
+        importer.import_plugin_intrinsic(cont, func);
 
-    // verify(world());
+    // verify(world_);
 
     todo_ |= importer.todo();
+
+    std::move(importer).finish(world_);
```

Comment:
The importer (any rewriter really) has finished it's job once you're done instantiating things in it, it doesn't need a concept of "finishing" and swapping the old world for the new one is a job for whoever calls a rewriter

Reply:
This is a pretty typical pattern with a builder like that. Requiring a move to get at the world inside the builder signifies that you're done with the object. And it gives the Importer a chance to perform some operations that need to be done once, after everything else is done. In my case, I needed to rewrite plugin intrinsics. That eventually moved elsewhere and wouldn't have been included here anyways, but I thought I'd leave this in for the future. Having client code mess directly with the internals of the `Importer` is IMHO very ugly.

Ideally, I would have simply had `.finish()` move the `world_` out of the `Importer`, but `World` isn't moveable for some reason (and copying a `World` apparently does something other than copy the `World`). Then you would just move-assign the new `World` to your old `World` instead of the swap hack. None of the cleanup code actually accesses any of the data members of `Importer`, so this would have allowed us to make those members private again (the `private` there was swapped for `public` as part of some hack 7 years ago with the commit message promising that this would be cleaned up later). Unfortunately, it turns out that the HLS stuff now really likes to mess with the `Importer`'s internals, so I had to revert that again.

Maybe it would be a good idea to just make the `World` moveable and replace all the swaps with proper moves? At least, it's unclear to me how swapping the `World` inside the `Importer` for your old `World` is anything other than a hack to work around the fact that `World` isn't movable? I didn't dare do this because I found the idea of changing the semantics of the `World` in such a fundamental way a bit too scary, given that I don't really understand the reasoning behind the current semantics. One problem with this swap hack to implement what should (?) be a move is that you can't rely on your old `World` being destroyed there and then as its lifetime instead is now bound to the lifetime of its new owner. This is unlikely to be a problem here, but I've had issues with this in other projects in the past…

Reply:
> You're conflating the concepts of rewriters and passes

To clarify: so did the old importer code by having `Importer` own the new world. The rewriter class in the `control` branch is just an interface to ease rewriting and avoids repeating some generic memoization code. See how other classes like `Mangler` and `CondEval` got aligned and merged with the generic `Rewriter`.

Reply:
> To clarify: so did the old importer code by having Importer own the new world

Yeah ok, I guess that's the part that confused me. I assumed that an `Importer` would be supposed to do the importing and produce a new `World`. I'll rebase this onto the development branch when I find some time.